### PR TITLE
packages/grafana-data/text: Improve escaping for special characters

### PR DESCRIPTION
**Pull request by joeblubaugh**
**What this PR does / why we need it**:

FilterInput escapes all input strings for special characters that might
be used in a RegExp by calling escapeStringForRegex, and using
unEscapeStringFromRegex for display. Both of these functions used
string.prototype.replace() for escaping. replace() only replaces the
first occurence of a search string unless called with a global RegExp.
The output of these functions was not necessarily safe to compile into a
RegExp literal.

This change creates RegExps for escapeStringForRegex and
unEscapeStringFromRegex to match all occurrences of the special
characters instead of just their first occurrence. This makes a variety
of strings safe for RegExp compilation.

**Special notes for your reviewer**:
Requesting specific reviews from @gillesdemey because this relates to a Grafana support escalation he's viewed, and from @joshhunt to give a look over my test style. Thanks for the pointers!

### CODE REVIEW

File: packages/grafana-data/src/text/string.test.ts [hunk 1]
```diff
@@ -57,12 +57,32 @@ describe('stringToMs', () => {
...
+    '[[[',
+    '[]',
+    '[(abc])',
+    'onetwothree',
+    '<namedgroup}(this is not a regex>',
+    'string\\with\\backslash',
+    'everyspecialchar([{])}.,/?&*^&<>#',
+  ])('should always produce output that compiles', (value) => {
+    const _ignored = new RegExp(escapeStringForRegex(value));
+    // There should be no error.
```
Comment (joshhunt) [inline packages/grafana-data/src/text/string.test.ts:85]: We can more fluently express our intention with `expect().not.toThrowError()`, which should give us a better failing test

```suggestion
    expect(() => new RegExp(escapeStringForRegex(value))).not.toThrowError()
```

https://jestjs.io/docs/expect#tothrowerror

**Review by joshhunt — COMMENTED** [review]
Great tests! 👍🏻

**Comment by gillesdemey** [comment]
Can confirm that it no longer throws on inputs like `[][` 👍

File: packages/grafana-data/src/text/string.ts [hunk 2]
```diff
@@ -1,20 +1,27 @@
...
 
+const specialCharEscape = new RegExp(specialMatcher, 'g');
+const specialCharUnescape = new RegExp('(\\\\)' + specialMatcher, 'g');
 export const escapeStringForRegex = (value: string) => {
   if (!value) {
     return value;
   }
 
-  return specialChars.reduce((escaped, currentChar) => escaped.replace(currentChar, '\\' + currentChar), value);
+  // For reviewers: this could simply be replaceAll if we're willing to
```
Comment (gillesdemey) [inline packages/grafana-data/src/text/string.ts:12]: I think `String.prototype.replaceAll()` throws if the regex doesn't have the `/g` modifier 🤔  Either way I don't believe we should update the compile target.

**Review by gillesdemey — APPROVED** [review]
LGTM 👍

**Pull request merged**